pkg/controller/worker/genericactuator/actuator_reconcile.go

```diff
@@ -195,7 +206,76 @@ func (a *genericActuator) Reconcile(ctx context.Context, worker *extensionsv1alp
 	return nil
 }
 
-func (a *genericActuator) deployMachineDeployments(ctx context.Context, cluster *controller.Cluster, worker *extensionsv1alpha1.Worker, existingMachineDeployments *machinev1alpha1.MachineDeploymentList, wantedMachineDeployments worker.MachineDeployments, classKind string, clusterAutoscalerUsed bool) error {
+func (a *genericActuator) addStateToMachineDeployment(ctx context.Context, worker *extensionsv1alpha1.Worker, wantedMachineDeployments workercontroller.MachineDeployments) error {
+	workerCopy := worker.DeepCopy()
+
+	if workerCopy.Status.State == nil || len(workerCopy.Status.State.Raw) <= 0 {
+		return nil
+	}
+	workerState := make(map[string]*workercontroller.MachineDeploymentState)
+
+	if err := json.Unmarshal(workerCopy.Status.State.Raw, &workerState); err != nil {
+		return err
+	}
+
+	for index, wantedMachineDeployment := range wantedMachineDeployments {
+		wantedMachineDeployments[index].State = workerState[wantedMachineDeployment.Name]
```

Comment:
Is my understanding correct that the only reason for this function being here is to preserve the state of already existing `MachineDeployment`s?

Reply:
This function is a part of restoration of the state of the desired `MachineDeployment`.
We do not restore state of something that already exist because this interferes the work of the MCM and gives out of date state.
The purpose of this function is to parse the Worker state to separate `MachineDeployment` states and assigns them to the appropriate `MachineDeployment`.
I know that this function can be optimised by setting the state to those `MachineDeployments` that does not exist but we have few `MachineDeployments` and we are assigning only pointers.

Reply:
Hm, I'm still confused, so this is already the code for restoring `MachineDeployment`s after a migration? Didn't the design doc tell that this is only triggered via a special "restore" annotation (cc @plkokanov)?

Reply:
I don't know what to say. Annotation operation is removed by the rapper we are using.
Unless we use some other kind of annotation I don't see how we can use the standard Reconcile flow and explicitly knowing that we are in migration.

Reply:
https://github.com/gardener/gardener/blob/master/docs/proposals/07-shoot-control-plane-migration.md states that `gardener.cloud/operation=restore` will be added in case Gardener wants the extensions to restore the state. It should not happen automatically "just because a machinedeployment is missing". This annotation should only be removed after the restoration succeeded. If the wrapper we are using removes this annotation then it should be adapted to onl remove the `gardener.cloud/operation=reconcile` and keep `gardener.cloud/operation=restore` until it can be safely removed.

Reply:
Sorry for the late reply. Indeed it should be done as it is stated in the GEP. I’m still testing if the current implementation works in all cases, however we cannot guarantee that it will be safe in the future (if something in the nvm changes) . Also what worries me and what I’m testing right now is that we have both the worker controller and mcm creating machines/machinesets. With the restore annotation the logic will be much simpler.

Reply:
An idea for the restore annotation: add a check for it to the OperationAnnotationWrapper.Reconcile and handle it just as the reconcile operation. Create a new class “inheriting” OperationAnnotationWrapper (which will be used by controllers perserving state) where the restore annotation is removed after successful reconciliation so that it can be used by the actuator to check whether to perform restoration.